### DISCUSSION TITLE
SAML Service Provider Metadata

### DISCUSSION THREAD

**Discussion by gabriel-hoff-vectorsolutions** [zitadel/zitadel#7286] [Q&A] [answered]
Hello!

I see that Zitadel offers an option to configure a SAML Identity Provider for an organization. However, when configuring Zitadel as a SAML-based Service Provider on an external IDP (e.g., JumpCloud), I'm prompted either for a SAML metadata file or URL.

I've tried searching Zitadel's documentation, but the only SAML metadata URL I could find is for configuring Zitadel as a SAML IDP instead of a SP: `$CUSTOM-DOMAIN/saml/v2/metadata` (found in https://zitadel.com/docs/apis/saml/endpoints#saml-20-metadata).

Is there any way to obtain the SAML metadata for using Zitadel as a Service Provider? Or is that metadata something that I have to generate myself?

**Comment by stebenz** [comment]
Hi @vs-gtadeu 
We will add to the documentation a guide on how to create a SAML provider on ZITADEL side.
The endpoint $CUSTOM-DOMAIN/saml/v2/metadata is if you use ZITADEL as IDP.

The metadata doesn't have to be generated and should be available after you create the SAML Identity provider in ZITADEL under the URL for example "https://example.zitadel.com/idps/<ID of the IDP>/saml/metadata .

**Reply by gabriel-hoff-vectorsolutions** [reply]
Thank you very much for the quick reply, @stebenz. I've tried the following endpoints, but all have redirected me back to `/ui/console/`:
- `/idps//saml/metadata`
- `/idps/saml/metadata`

Are any of them the correct ones? Also, should the Identity Provider ID or Organization ID be passed as a parameter somehow?

**Reply by stebenz** [reply]
/idps/251026351862061042/saml/metadata for example would have been correct, which is the identity provider ID.
The organization ID is included through the domain.

But let me look into this and I will come back to you ASAP.

**Reply by gabriel-hoff-vectorsolutions** [reply]
Okay. I've tried with `/idps/251016397537155057/saml/metadata`, and I'm still getting redirected to `/ui/console`.

**Reply by gabriel-hoff-vectorsolutions** [reply]
@stebenz any news on this?

**Reply by stebenz** [reply]
Yes, it's with our ingress, I already tested it successfully on our QA system, and will try to get it out as fast as possible and I can tell you more tomorrow.

**Reply by stebenz** [reply]
@vs-gtadeu Could you please try it again with the path I mentioned before `/idps/251016397537155057/saml/metadata`?

**Reply by john-jorgensen-vectorsolutions** [reply]
Hi @stebenz, I work with @vs-gtadeu and have also been trying out using Zitadel as a SP with an external SAML IdP. The `/idps/{idpID}/saml/metadata` endpoint works now, thank you!

If you wouldn't mind answering one more question, I'd like to ask about the login flow with a SAML IdP: After clicking the button on the Zitadel login page to log in with my SAML IdP, I'm being directed to a blank page with a submit button:
<img width="1024" alt="Screenshot 2024-02-01 at 12 20 25 PM" src="https://github.com/zitadel/zitadel/assets/99289501/2da00159-84d4-48ca-b041-0b14620741f2">
Is there a way to have the external IdP login button take the user directly to the external IdP login, without this submit button page?

**Reply by gabriel-hoff-vectorsolutions** [reply]
Looks like there is JavaScript error on that page:

```
Refused to execute inline script because it violates the following Content Security Policy directive: "script-src 'self' 'nonce-LOcdf8Gx/DOnC3nJRx7frXsYaVrLFLDPvipFs9sING8='". Either the 'unsafe-inline' keyword, a hash ('sha256-AjPdJSbZmeWHnEc5ykvJFay8FTWeTeRbs9dutfZ0HqE='), or a nonce ('nonce-...') is required to enable inline execution.
```

![image](https://github.com/zitadel/zitadel/assets/62257776/5c420fac-eee0-481a-b86c-33c3a32ceafa)

**Reply by stebenz** [reply]
With what binding do you connect your SAML IDP? Post or Redirect-binding?
Maybe you can change this, otherwise for the CSP I will have a look at it, as I have not yet encountered that problem.

**Reply by john-jorgensen-vectorsolutions** [reply]
I've tried using the Post and Redirect bindings, both get stopped at that page because the inline javascript isn't being run

**Reply by livio-a** [reply]
We're currently working on a fix for the CSP issue. This will also make sure that the configured binding is used.
https://github.com/zitadel/zitadel/pull/7341

This will be released and deployed to zitadel.cloud within a day.

**Reply by livio-a** [reply, marked as answer]
It's just been deployed to zitadel.cloud. Please let us know if it works for you as well.

**Reply by john-jorgensen-vectorsolutions** [reply]
Both bug fixes work! I'm able to log in with post binding without getting stuck on the SAML post page, and changing the binding to redirect doesn't send me to the SAML post page.

Thank you for identifying and fixing those bugs so quickly!
